docs/proposals/deletions_object_storage.md

```diff
@@ -0,0 +1,100 @@
+
+*   **Accidental insertion of confidential data:** This is a scenario where a user accidentally inserts confidential data and wants to delete it immediately. In this case, the user expects their request to accelerate an immediate deletion of the series pertaining to the blocks concerned with that specific data to be deleted.
+*   **GDPR:** Masking data and eventual deletion is expected.
+*   **Deletions to sustain user requirements:** Let’s assume the user has some data which leads to some unexpected results or causes performance degradation (due to high cardinality) and the user wants to restore the previous data set-up for obtaining the desired results. In this scenario, the user would want to send a request to mask the data for the time being as there isn’t a high priority requirement to delete the data but eventually during the compaction the user can expect the data to be deleted by the compactor not leading to any major performance issues.
+*   **Helps achieving series based retention (e.g rule aggregation) [#903](https://github.com/thanos-io/thanos/issues/903).**
+
+## Goals
+
+*   Unblock users and allow series deletion in the object storage using tombstones.
+*   Deal with compaction and downsampling of blocks with tombstones.
+*   Performing deletions at admin level.
+
+## Proposed Approach
+
+*   We propose to implement deletions via the tombstones approach using a CLI tool.
+*   A tombstone is proposed to be a **Custom format global - single file per request**.
+*   **Why Custom format global - single file per request**? :
+    *   https://cloud-native.slack.com/archives/CL25937SP/p1595954850430300
+    *   We can easily have multiple writers
+    *   No need to have index data of the blocks
+*   A user is expected to enter the following details for performing deletions:
+    *   **label matchers**
+    *   **start timestamp**
+    *   **end timestamp** (start and end timestamps of the series data the user expects to be deleted)
+*   The entered details are processed by the CLI tool to create a tombstone file (unique for a request), and then the tombstone file is uploaded to the object storage making it accessible to all components.
+*   Store Gateway masks the series on processing the global tombstone files from the object storage.
+*   During compaction or downsampling, we check if the blocks being considered have series corresponding to a tombstone file, if so we delete the data and continue with the process.
+
+## Open Questions
+*   Optimization for file names?
+*   Do we create a tombstone irrespective of the presence of the series with the given matcher? Or should we check for the existence of series that corresponds to the deletion request?
+*   Should we consider having a threshold for compaction or downsampling where we check all the tombstones and calculate the percentage of deletions?
+
+## Considerations
+
+*   The old blocks and the tombstones are deleted during compaction time.
```

Comment:
I think this opens to a race condition.

Above it's mentioned this:
> Store Gateway masks the series on processing the global tombstone files from the object storage.

If the store gateway scans deleted tombstones before the old blocks are effectively offloaded by the store gateway (we do soft delete first, apply a delay and then delete hard delete them) we may open to a time window during which the tombstone is deleted, the store gateway removed it from its local map, the old block is still queried and we do re-expose deleted series.

Reply:
Yes, it has to be delayed-deleted or something else... It's not easy, because:

* How we know there are no more stuff to delete then in mentioned block?
* What if someone backfills block that touches deleted period?
* What if min-time is .. not specified? Does it has to be?


Behaviour on all those questions have to be defined in this proposal, explicit and simple to understand for user of deletion API I think ):